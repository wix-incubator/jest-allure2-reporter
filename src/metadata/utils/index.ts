export * from './getStage';
export * from './getStart';
export * from './getStatusAndDetails';
export * from './getStop';
