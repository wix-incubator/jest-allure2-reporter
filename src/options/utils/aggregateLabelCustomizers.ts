/* eslint-disable unicorn/no-array-reduce */
import type {
  Extractor,
  ExtractorContext,
  TestFileCustomizer,
  TestCaseCustomizer,
} from 'jest-allure2-reporter';
import type { Label } from 'jest-allure2-reporter';

import { asExtractor } from './asExtractor';

type Customizer = TestFileCustomizer | TestCaseCustomizer;
export function aggregateLabelCustomizers<C extends Customizer>(
  labels: C['labels'] | undefined,
): Extractor<Label[]> | undefined {
  if (!labels || typeof labels === 'function') {
    return labels as Extractor<Label[]> | undefined;
  }

  const extractors = Object.keys(labels).reduce(
    (accumulator, key) => {
      const extractor = asExtractor(labels[key]) as Extractor<string[]>;
      if (extractor) {
        accumulator[key] = extractor;
      }
      return accumulator;
    },
    {} as Record<string, Extractor<string[]>>,
  );

  const names = Object.keys(extractors);

  return (context: ExtractorContext<Label[]>): Label[] | undefined => {
    const other: Label[] = [];
    const found = names.reduce(
      (found, key) => {
        found[key] = [];
        return found;
      },
      {} as Record<string, string[]>,
    );

    if (context.value) {
      for (const label of context.value) {
        if (found[label.name]) {
          found[label.name].push(label.value);
        } else {
          other.push(label);
        }
      }
    }

    const result = [
      ...other,
      ...names.flatMap((name) => {
        const extractor = extractors[name];
        const aContext: ExtractorContext<string[]> = {
          ...context,
          value: asArray(found[name]),
        };
        const value = asArray(extractor(aContext));
        return value ? value.map((value) => ({ name, value }) as Label) : [];
      }),
    ];

    return result;
  };
}

function asArray<T extends string>(
  value: T | T[] | undefined,
): T[] | undefined {
  if (Array.isArray(value)) {
    return value.length > 0 ? value : undefined;
  } else {
    return value ? [value] : [];
  }
}
