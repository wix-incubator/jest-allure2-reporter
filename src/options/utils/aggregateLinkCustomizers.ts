import type { Link } from '@noomorph/allure-js-commons';
import type {
  Extractor,
  ExtractorContext,
  TestFileCustomizer,
  TestCaseCustomizer,
} from 'jest-allure2-reporter';

type Customizer = TestFileCustomizer | TestCaseCustomizer;

export function aggregateLinkCustomizers<C extends Customizer>(
  links: C['links'] | undefined,
): Extractor<Link[]> | undefined {
  if (!links || typeof links === 'function') {
    return links as Extractor<Link[]> | undefined;
  }

  return (context: ExtractorContext<Link[]>) => {
    return context.value
      ?.map((link) => {
        const extractor = links[link.type ?? ''];
        return extractor ? extractor({ ...context, value: link } as any) : link;
      })
      ?.filter(Boolean) as Link[] | undefined;
  };
}
