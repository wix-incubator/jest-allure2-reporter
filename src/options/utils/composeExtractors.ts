import type { Extractor, ExtractorContext } from 'jest-allure2-reporter';

export function composeExtractors<T, C extends ExtractorContext<T>>(
  a: Extractor<T, C> | undefined,
  b: Extractor<T, C>,
): Extractor<T, C> {
  return a ? (context) => a({ ...context, value: b(context) }) : b;
}
