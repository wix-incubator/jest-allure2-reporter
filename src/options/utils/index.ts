export * from './aggregateLabelCustomizers';
export * from './aggregateLinkCustomizers';
export * from './asExtractor';
export * from './composeExtractors';
export * from './resolvePlugins';
export * from './stripStatusDetails';
