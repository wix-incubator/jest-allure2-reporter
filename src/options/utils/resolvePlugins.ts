import type {
  Plugin,
  PluginConstructor,
  PluginReference,
  PluginDeclaration,
  PluginContext,
} from 'jest-allure2-reporter';

export function resolvePlugins(
  context: PluginContext,
  plugins: PluginDeclaration[] | undefined,
): Promise<Plugin[]> {
  if (!plugins) {
    return Promise.resolve([]);
  }

  const promises = plugins.map((plugin) => {
    return Array.isArray(plugin)
      ? resolvePlugin(context, plugin[0], plugin[1])
      : resolvePlugin(context, plugin, {});
  });

  return Promise.all(promises);
}

async function resolvePlugin(
  context: PluginContext,
  reference: PluginReference,
  options: Record<string, unknown>,
): Promise<Plugin> {
  let createPlugin: PluginConstructor;

  if (typeof reference === 'string') {
    const rootDirectory = context.globalConfig.rootDir;
    const resolved = require.resolve(reference, { paths: [rootDirectory] });
    createPlugin = await import(resolved);
  } else {
    createPlugin = reference;
  }

  return createPlugin(options, context);
}
