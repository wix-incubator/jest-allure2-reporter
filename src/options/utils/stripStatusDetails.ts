import stripAnsi from 'strip-ansi';
import type { StatusDetails } from 'jest-allure2-reporter';

export function stripStatusDetails(
  statusDetails?: StatusDetails,
): StatusDetails | undefined {
  if (!statusDetails) {
    return undefined;
  }

  const { message, trace } = statusDetails;
  return {
    message: message ? stripAnsi(message) : undefined,
    trace: trace ? stripAnsi(trace) : undefined,
  };
}
