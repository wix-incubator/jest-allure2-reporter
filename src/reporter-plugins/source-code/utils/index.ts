export * from './extractTypescriptAST';
export * from './extractTypeScriptCode';
export * from './importTypeScript';
export * from './FileNavigatorCache';
export * from './LineNavigator';
