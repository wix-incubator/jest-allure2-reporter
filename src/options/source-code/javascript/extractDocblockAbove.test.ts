import { FileNavigator } from '../../../utils';

import { extractDocblockAbove as extractJsDocument_ } from './extractDocblockAbove';

const FIXTURES = [
  `\
/**
 * This is a multiline docblock
 */`,
  '  /** This is a single line docblock */',
  `/**
 * This is a broken docblock
   * but it's still a docblock
*/`,
  `/**
    * This is a weird two-line docblock */`,
  `/*
    * This is not a docblock
    */`,
].map((sourceCode) => sourceCode + '\n' + 'function test() {}\n');

describe('extractJsDoc', () => {
  const extract = (index: number, line: number) =>
    extractJsDocument_(new FileNavigator(FIXTURES[index]), line);

  it('should extract a multiline docblock', () => expect(extract(0, 4)).toMatchSnapshot());

  it('should extract a single line docblock', () => expect(extract(1, 2)).toMatchSnapshot());

  it('should extract a broken docblock', () => expect(extract(2, 5)).toMatchSnapshot());

  it('should extract a weird two-line docblock', () => expect(extract(3, 3)).toMatchSnapshot());

  it('should not extract a non-docblock', () => expect(extract(4, 4)).toMatchSnapshot());

  it('should ignore out of range line index', () => expect(extract(0, 5)).toBe(''));

  it('should ignore zero line index', () => expect(extract(0, 0)).toBe(''));

  it('should ignore the middle of a docblock', () => expect(extract(0, 2)).toBe(''));
});
