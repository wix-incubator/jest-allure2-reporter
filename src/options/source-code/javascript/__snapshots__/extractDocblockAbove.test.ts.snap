// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractJsDoc should extract a broken docblock 1`] = `
"/**
 * This is a broken docblock
   * but it's still a docblock
*/"
`;

exports[`extractJsDoc should extract a multiline docblock 1`] = `
"/**
 * This is a multiline docblock
 */"
`;

exports[`extractJsDoc should extract a single line docblock 1`] = `"  /** This is a single line docblock */"`;

exports[`extractJsDoc should extract a weird two-line docblock 1`] = `
"/**
    * This is a weird two-line docblock */"
`;

exports[`extractJsDoc should not extract a non-docblock 1`] = `""`;
